content/blog/2020-12-18-type-and-effect-system.md

```diff
@@ -0,0 +1,93 @@
+
+The standard library provides a mapping from function names to function types, which is used as the starting type context for typechecking. This had the unfortunate constraint of disallowing `print` from supporting a variable number of arguments; varargs support would require an extra wrapper around types in the type context or a new `Any` type, neither of which were appealing options for this project.
+
+## Interpreter
+The interpreter takes a parsed program AST as input and executes the program, relying on the typechecker as a preprocessing step to better enforce correctness. Starting from an empty map from variable names to their values, each statement of the list of top-level statements comprising the entire program is executed in order, with the resulting context being threaded from one statement to the next. 
+
+The most difficult part of this module was correctly evaluating lambdas and closing around the variables needed, especially in partial application situations. To that end, `Lambda (x, t, s)` had to be extended in the code to `Lambda (x, t, ctx, s)`, with `ctx` representing the context of the closure upon creation. Then, partial application could be implemented as nested lambda expressions, and evaluating one layer of a lambda stack would encode its argument into the context of the next layer. 
+
+## Optimizer
+The optimizer included with the language is very simple, demonstrating only one of the possible optimizations. In essence, relying on the typechecker to properly enforce the effect annotations, the optimizer is able to collapse pure expressions that have fixed inputs, essentially performing a combination of inlining and constant propagation. This can be seen in the `fact.txt` sample program, which automatically precomputes the factorial of 5 after an optimization pass. A further dead code elimination pass, not implemented here, could drastically reduce program size in this case. 
+
+For each of the small benchmark programs in the `test/basic/` subdirectory, an optimized version will run in the `test/optimize/` subdirectory. The comparison script `test/compare.sh` will diff the results from these two runs; at the time of writing, the results of this diff are as follows:
+- `fact.txt` is improved from 24 executed statements to 4
+- `func.txt` is improved from 8 executed statements to 6
+- `func3.txt` is improved from 20 executed statements to 8
+- all other program execution counts were unchanged
+
+Though not deep, this level of empirical evaluation seemed appropriate considering the scope of the optimization: any memory manipulation, output, or input (e.g. variable input) would result in no optimizations at all.
```

Comment:
Always use a comma after "e.g." https://capra.cs.cornell.edu/styleguide/#egie
(This also occurs again below.)